### PR TITLE
enable prompt tuning/prefix tuning/p tuning clm and example

### PR DESCRIPTION
**Pull request by sywangyi**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sywangyi** [comment]
@libinta @regisss  please help review. prompt tunning is just a first step, next I will enable prefix-tuning and ptuning based on the example

**Comment by sywangyi** [comment]
prompt tunning, prefix-tuning and ptuning is enabled for llama finetune and inference

**Comment by sywangyi** [comment]
inference part needs model change to adapter prompt tuning/prefix tuning. since virtual token is used in prompt/p tuning and past_key_value is extended in prefix tuning.

**Comment by sywangyi** [comment]
also upgrade the peft to latest tag 0.9.0

**Comment by sywangyi** [comment]
prompt tunning+deepspeed zero3 issue fixed by https://github.com/huggingface/peft/pull/1591

**Comment by sywangyi** [comment]
find deepspeed zero3+prompt tunning hang after saving checkpoint. fix by https://github.com/huggingface/transformers/pull/29980. will port similar change to optimum habana trainer.

File: examples/language-modeling/run_prompt_tuning_clm.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
+import logging
```
Comment (libinta) [inline examples/language-modeling/run_prompt_tuning_clm.py:1]: please add the license and source of this file

File: examples/language-modeling/README.md [hunk 2]
```diff
@@ -636,6 +636,64 @@ DEEPSPEED_HPU_ZERO3_SYNC_MARK_STEP_REQUIRED=1 LOWER_LIST=ops_bf16.txt python3 ..
...
+Default peft_type is prompt_tuning, you could enable prefix-tuning or p-tuning using "--peft_type prefix_tuning" or "--peft_type p_tuning"
+
+use the prompt finetuned model for text-generation:
+```bash
+python3 ../text-generation/run_generation.py \
+    --model_name_or_path meta-llama/Llama-2-7b-hf  \
+    --max_new_tokens 128 \
+    --bf16 \
+    --use_kv_cache \
+    --batch_size 1 \
```
Comment (libinta) [inline examples/language-modeling/README.md:689]: is there any reason to just use batch_size 1?

File: examples/language-modeling/run_prompt_tuning_clm.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+            encoder_hidden_size=model_args.encoder_hidden_size,
+        )
+    elif model_args.peft_type == "prefix_tuning":
+        peft_config = PrefixTuningConfig(
+            task_type=TaskType.CAUSAL_LM,
+            num_virtual_tokens=model_args.num_virtual_tokens,
+        )
+
+    text_column = "Tweet text"
+    label_column = "text_label"
```
Comment (libinta) [inline examples/language-modeling/run_prompt_tuning_clm.py:230]: are these 2 specific to the dataset in-use? if so, can we check dataset name and put others as unimplemented?

File: examples/language-modeling/README.md (hunk 2 above)
Comment (sywangyi) [reply examples/language-modeling/README.md:689]: the train set only has 50 elements

File: examples/language-modeling/run_prompt_tuning_clm.py (hunk 3 above)
Comment (sywangyi) [reply examples/language-modeling/run_prompt_tuning_clm.py:230]: yes

File: optimum/habana/peft/peft_model.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            if model_kwargs["past_key_values"][0][0].shape[-2] >= model_kwargs["input_ids"].shape[1]:
+                model_kwargs["input_ids"] = model_kwargs["input_ids"][:, -1:]
+
+        if model_kwargs.get("attention_mask", None) is not None:
+            size = model_kwargs["input_ids"].shape[0], peft_config.num_virtual_tokens
+            prefix_attention_mask = torch.ones(size).to(model_kwargs["input_ids"].device)
+            model_kwargs["attention_mask"] = torch.cat((prefix_attention_mask, model_kwargs["attention_mask"]), dim=1)
+            token_idx = model_kwargs.get("token_idx", None)
+            if token_idx is not None:
+                token_idx = token_idx + peft_config.num_virtual_tokens
```
Comment (libinta) [inline optimum/habana/peft/peft_model.py:45]: did you check if token_idx_cpu is set for this? if it is, it needs to be updated,right?

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -354,7 +353,7 @@ def pre_attn_forward(
 
         kv_seq_len = key_states.shape[-2]
         if past_key_value is not None:
-            if token_idx is None:
+            if token_idx is None or past_key_value[0].shape[-2] < token_idx:
                 if hasattr(past_key_value, "get_usable_length"):
```
Comment (libinta) [inline optimum/habana/transformers/models/llama/modeling_llama.py:357]: Why do we need this change? past_key_value[0].shape[-2] < token_idx:

Comment (sywangyi) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: thanks for the remind, I update the PR to support reuse_cache, bucket_size optimization in prompt/ptuning/prefix ptuning generation. past_key_value[0].shape[-2] < token_idx. this is the case for prefix tuning peft text-generation. which extra pask_key_value will be attached to generate the 1st token.

**Comment by regisss** [comment]
@sywangyi I just merged #809, can you merge it into this branch and solve the merge conflicts please?

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 6]
```diff
@@ -364,17 +364,24 @@ def pre_attn_forward(
...
                     kv_seq_len += past_key_value[0].shape[-2]
             else:
-                if reuse_cache:
+                if reuse_cache and not isinstance(past_key_value[0], torch.Tensor):
                     kv_seq_len = past_key_value[0][-2]
                 else:
-                    kv_seq_len = past_key_value[0].shape[-2]
+                    kv_seq_len = (
+                        past_key_value[0].shape[-2] + kv_seq_len
+                        if past_key_value[0].shape[-2] < token_idx.cpu()
```
Comment (libinta) [inline optimum/habana/transformers/models/llama/modeling_llama.py:372]: this could cause tgi performance issue as token_idx needs to be moved to cpu each time. I would suggest to have model argument

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 7]
```diff
@@ -385,8 +392,15 @@ def pre_attn_forward(
                         key_states.shape, dtype=self.k_proj.weight.dtype, device=key_states.device
                     )
                     past_key_value = (past_key, past_value)
-                key_states = self.k_cache.update(past_key_value[0], key_states, 2, token_idx, self.inp_seq_len)
-                value_states = self.v_cache.update(past_key_value[1], value_states, 2, token_idx, self.inp_seq_len)
+                if token_idx is not None and token_idx.cpu() > past_key_value[0].shape[2]:
```
Comment (libinta) [inline optimum/habana/transformers/models/llama/modeling_llama.py:395]: prefer not to have .cpu call for performance purpose

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 8]
```diff
@@ -727,7 +751,10 @@ def forward(
 
         if past_key_values is not None and use_cache:  # kept for BC (cache positions)
             if reuse_cache:
-                past_seen_tokens = past_key_values[0][0][2]
+                if isinstance(past_key_values[0][0], torch.Tensor):
```
Comment (libinta) [inline optimum/habana/transformers/models/llama/modeling_llama.py:754]: can you use the num_virtual_token to check here?

File: examples/text-generation/run_generation.py [hunk 9]
```diff
@@ -247,6 +247,11 @@ def setup_parser(parser):
         action="store_true",
         help="Whether to use torch compiled model or not.",
     )
+    parser.add_argument(
+        "--no_ignore_eos",
```
Comment (regisss) [inline examples/text-generation/run_generation.py:251]:
```suggestion
        "--ignore_eos",
```

File: examples/text-generation/run_generation.py [hunk 10]
```diff
@@ -376,6 +381,7 @@ def generate(size=None, reduce_recompile=False):
                 hpu_graphs=args.use_hpu_graphs,
                 profiling_steps=args.profiling_steps,
                 profiling_warmup_steps=args.profiling_warmup_steps,
+                ignore_eos=not args.no_ignore_eos,
```
Comment (regisss) [inline examples/text-generation/run_generation.py:384]:
```suggestion
                ignore_eos=args.no_ignore_eos,
```

File: examples/text-generation/run_generation.py [hunk 11]
```diff
@@ -556,6 +562,7 @@ def generate_dataset(batch):
...
```
Comment (regisss) [inline examples/text-generation/run_generation.py:565]:
```suggestion
                ignore_eos=args.no_ignore_eos,
```

File: examples/text-generation/utils.py [hunk 12]
```diff
@@ -285,11 +289,17 @@ def peft_model(args, model_dtype, logger, **model_kwargs):
 
         model = AutoModelForCausalLM.from_pretrained(args.model_name_or_path, torch_dtype=model_dtype, **model_kwargs)
         model = PeftModel.from_pretrained(model, args.peft_model, torch_dtype=model_dtype, **model_kwargs)
+    if hasattr(model, "merge_and_unload"):
```
Comment (regisss) [inline examples/text-generation/utils.py:292]: Do you have an example of model that doesn't have the `merge_and_unload` method?

**Review by libinta — COMMENTED** [review]
@sywangyi can you add ci test for this?

File: examples/text-generation/utils.py (hunk 12 above)
Comment (sywangyi) [reply examples/text-generation/utils.py:292]: prompt tuning/p tuning and prefix tuning do not have the merged_and_unload method

**Review by libinta — APPROVED** [review]

**Comment by regisss** [comment]
@sywangyi Why calling the new arg `--no_ignore_eos` and not simply `--ignore_eos` as everywhere else in the codebase? If you just want to have it true by default, I think this arg should rather be a bool with the default value True. It's clearer than an arg "no_ignore_eos" with `store_false`.

**Comment by sywangyi** [comment]
@regisss  arg bool does not work, see https://stackoverflow.com/questions/15008758/parsing-boolean-values-with-argparse. anyway, I picked up one solution to fix it.

**Comment by regisss** [comment]
I pushed a new commit where the logic is simpler. It only works with Python 3.9+, which is fine as Habana Docker images don't support Ubuntu 20.04 anymore.
